### PR TITLE
Testing Python Script for Lighthouse and DataDog integration

### PR DESCRIPTION
**Pull request by ardelato**
WIP

### CODE REVIEW

**Comment by ardelato** [comment]
image

https://app.datadoghq.com/dashboard/d23-itu-k3m/lighthouse-reports-prototype?from_ts=1685552988792&to_ts=1685556588792&live=true

**Comment by ardelato** [comment]
@andyg0808 I didn't port the entirety of the ruby scripts as the core functionality was very basic.

Let me know if this is what you had in mind.

File: lib/lighthouse.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    def run(self, url):
+        command = self.command + [url]
+        result = subprocess.run(command, stdout=subprocess.PIPE)
+
+        if result.returncode != 0:
+            raise Exception('Lighthouse failed')
+
+        json_results = json.loads(result.stdout)
+        metrics = self.parse_json_report(json_results)
+        self.send_metrics_to_datadog(url, metrics)
```
Comment (andyg0808) [inline lib/lighthouse.py:18]: I'd recommend these lines be in a function in `main.py`, simply as a separation of concerns measure: I think the code will be easier to work with if `lighthouse.py` only interfaces with `lighthouse` and `main.py` does the stitching-together of `lighthouse.py` and `datadog.py`.

File: lib/datadog.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from datadog_api_client.api_client import ApiClient
+from datadog_api_client.v2.api import metrics_api
+from datadog_api_client.v2.model.metric_point import MetricPoint
+from datadog_api_client.v2.model.metric_series import MetricSeries
+from datadog_api_client.v2.model.metric_payload import MetricPayload
+from datadog_api_client.v2.model.metric_intake_type import MetricIntakeType
+import os, time
+
+from dotenv import load_dotenv
+load_dotenv()
```
Comment (andyg0808) [inline lib/datadog.py:11]: I'd recommend doing this in `main.py`, not here.

Generally, it's easier to work with code if global/external side-effects are kept to a limited set of files and libraries are kept purely focused on one task.

Comment (ardelato) [reply lib/datadog.py]: Addressed in [`6ee75f1` (#13)](https://github.com/iFixit/lighthouse-docker/pull/13/commits/6ee75f1ab4680278807a72dde0bc2a067be55a62)

**Comment by ardelato** [comment]
@andyg0808 besides extracting some of the functions in `main.py` do you think this is sufficient?

File: lib/main.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    page_path = urlparse(url).path.split('/')[1:]
+    if page_path:
+        page_type = page_path[0]
+    else:
+        page_type = 'Home'
+
+    tags = {
+        'page_type': page_type,
+        'url': url,
+        'test': 'test'
```
Comment (andyg0808) [inline lib/main.py:51]: Is `test` supposed to be in here?

File: lib/main.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+            print(f'Sending metrics to datadog for {url}')
+            send_metrics_to_datadog(url, metrics)
+
+            print(f'Finished sending metrics to datadog for {url}\n')
+
+            print('=' * 80)
+        except Exception as e:
+            print(f'Failed to run lighthouse for {url}: {e}')
+            raise e
```
Comment (andyg0808) [inline lib/main.py:80]: I'd suggest throwing a new exception with the `print` value as its message; if I'm understanding the docs correctly, I think that'll also include `e` as `__context__` on the new exception, and not using `print` means the error message doesn't end up on stdout.

**Review by andyg0808 — APPROVED** [review]
CR :+1: 
A couple comments, but looks reasonably good to me.

File: lib/main.py (hunk 3 above)
Comment (ardelato) [reply lib/main.py:51]: Yes, its to be able to filter out any metrics sent to DataDog when querying. 

I should note this is just a draft pull, I won't actually commit the changes as is.

File: lib/datadog_api.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    @staticmethod
+    def get_instance():
+        if DataDogApiClient.__instance == None:
+            return DataDogApiClient()
+        return DataDogApiClient.__instance
+
+    def __init__(self):
+        """ Virtually private constructor. """
+        if DataDogApiClient.__instance != None:
+            raise Exception("This class is a singleton!")
```
[the remark this answers is not in the document]
Comment (ardelato) [reply lib/datadog_api.py:16]: No real reason, I thought that we would really only need one DataDog Api Client so I thought of using a singleton. 

I'll remove it in the final pull.

[the remark this answers is not in the document]
Comment (andyg0808) [reply lib/datadog_api.py:16]: :+1: 
Just as a general software development principle, it turns out it's better to avoid singletons.

**Pull request closed without merging**